packages/terser/src/index.js

```diff
@@ -43,31 +57,91 @@ function terserDirectory(input) {
+
+                log_error(
+                    'errored: ' + inputFile + '\n', 'OUT: ' + data.out + '\n',
+                    ' ERR: ' + data.err + '\n', ' code:' + data.code);
+              } else {
+                log_verbose('finished: ', inputFile);
+              }
+              --active;
+              next();
+            },
+            (err) => {
+              --active;
+              log_error('errored: [spawn exception]', inputFile, '\n' + err)
+              errors.push(inputFile)
+              next();
+            })
+  }
+
+  function next() {
+    if (work.length) {
+      exec(work.shift());
+    } else if (!active) {
+      if (errors.length) {
+        log_error('terser errored processing javascript in directory.')
+        process.exitCode = 2;
+      }
+      // NOTE: work is done at this point and node should exit here.
+    }
+  }
+
   fs.readdirSync(input).forEach(f => {
     if (f.endsWith('.js')) {
       const inputFile = path.join(input, path.basename(f));
       const outputFile = path.join(output, path.basename(f));
-      // We don't want to implement a command-line parser for terser
-      // so we invoke its CLI as child processes, just altering the input/output
-      // arguments. See discussion: https://github.com/bazelbuild/rules_nodejs/issues/822
-      // FIXME: this causes unlimited concurrency, which will definitely eat all the RAM on your
-      // machine;
-      //        we need to limit the concurrency with something like the p-limit package.
-      // TODO: under Node 12 it should use the worker threads API to saturate all local cores
-      child_process.fork(
-          // Note that the fork API doesn't do any module resolution.
-          require.resolve('terser/bin/uglifyjs'), [inputFile, '--output', outputFile, ...residual]);
+
+      if (active < TERSER_CONCURENCY) {
+        exec([inputFile, outputFile]);
+      } else {
+        work.push([inputFile, outputFile])
+      }
     }
   });
 }
 
-if (!inputs.find(isDirectory)) {
+if (!inputs.find(isDirectory) && inputs.length) {
   // Inputs were only files
   // Just use terser CLI exactly as it works outside bazel
-  require('terser/bin/uglifyjs');
+  require(TERSER_BINARY || 'terser/bin/uglifyjs');
+
 } else if (inputs.length > 1) {
   // We don't know how to merge multiple input dirs to one output dir
   throw new Error('terser_minified only allows a single input when minifying a directory');
-} else {
+
+} else if (inputs[0]) {
   terserDirectory(inputs[0]);
 }
+
+function spawn(cmd, args) {
+  return new Promise((resolve, reject) => {
+    const err = [];
+    const out = [];
+    let proc = child_process.spawn(cmd, args);
+
+    proc.stdout.on('data', (buf) => {
+      out.push(buf);
+    });
+    proc.stderr.on('data', (buf) => {err.push(buf)})
+    proc.on('exit', (code) => {
+      resolve({out: Buffer.concat(out), err: err.length ? Buffer.concat(err) : false, code: code});
```

Comment:
I'm a bit surprised we never reject

Reply:
failures are an expected output not an exceptional one so we just transfer the data reliably